app/http.php

```diff
@@ -83,21 +84,67 @@
 App::setResource('logger', fn () => $register->get('logger'));
 App::setResource('algorithm', fn () => $register->get('algorithm'));
 
-// Balancing must NOT be registry. This has to run on every request
-App::setResource('balancing', function (Table $state, Algorithm $algorithm) {
-    $balancing = new Balancing($algorithm);
+// Balancer must NOT be registry. This has to run on every request
+App::setResource('balancerGroup', function (Table $state, Algorithm $algorithm, Request $request) {
+    $runtimeId = $request->getHeader('x-opr-runtime-id', '');
 
-    $balancing->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
+    $group = new Group();
+
+    // Cold-started-only options
+    $balancer1 = new Balancer($algorithm);
+
+    // Only online executors
+    $balancer1->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
+
+    // Only low host-cpu usage
+    $balancer1->addFilter(function ($option) {
+        /**
+         * @var array<string,mixed> $state
+         */
+        $state = \json_decode($option->getState('state', '{}'), true);
+        return ($state['usage'] ?? 0) < 80;
+    });
+
+    // Only low runtime-cpu usage
+    if ($runtimeId !== '') {
+        $balancer1->addFilter(function ($option) use ($runtimeId) {
+            /**
+             * @var array<string,mixed> $state
+             */
+            $state = \json_decode($option->getState('state', '{}'), true);
+
+            /**
+             * @var array<string,mixed> $runtimes
+             */
+            $runtimes = $state['runtimes'];
+
+            /**
+             * @var array<string,mixed> $runtime
+             */
+            $runtime = $runtimes[$runtimeId] ?? [];
+
+            return ($runtime['usage'] ?? 0) < 80;
+        });
+    }
+
+    // Any options
+    $balancer2 = new Balancer($algorithm);
+    $balancer2->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
 
     foreach ($state as $state) {
```

Comment:
reminder to change the name of the loop variable

Reply:
Changed in diffetent PR: https://github.com/open-runtimes/proxy/pull/7